wire-schema/src/main/java/com/squareup/wire/schema/Schema.java

```diff
@@ -125,12 +132,19 @@ static Schema fromFiles(Iterable<ProtoFile> sourceProtoFiles, Loader pathFilesLo
     return new Linker(pathFilesLoader).link(sourceProtoFiles);
   }
 
-  private static ImmutableMap<String, Type> buildTypesIndex(Iterable<ProtoFile> protoFiles) {
+  private static ImmutableMap<String, Type> buildTypesIndex(Iterable<ProtoFile> protoFiles,
+      Map<ProtoType, ProtoFile> protoFilesIndex) {
     Map<String, Type> result = new LinkedHashMap<>();
     for (ProtoFile protoFile : protoFiles) {
       for (Type type : protoFile.getTypes()) {
         index(result, type);
       }
+      for (Type type : result.values()) {
```

Comment:
this loop should be out outside the enclosing for loop

Reply:
I don’t think this is the best way to index these

Reply:
... instead, make the `index()` method take the `protoFilesIndex` ?

Reply:
How could I get my hands on the right protoFile if I'm outside the enclosing loop?

Reply:
I can pass both the protoFile and the `protoFilesIndex` to `index()`